File: two pointers/lc82.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
    
        dummy = ListNode()
        dummy.next = head

        pre = dummy
        cur = head
        
        while cur is not None:
            if cur.next is not None and cur.val == cur.next.val:
                while cur.next is not None and cur.val == cur.next.val:
                    cur = cur.next
                pre.next = cur.next
            else:
                pre = cur
            cur = cur.next
        
        return dummy.next

File: phase 1/01 - two pointers/lc82.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    
    if (head == null || head.next ==null){return head}

    let dummy = new ListNode()
    dummy.next = head

    let pre = dummy
    let cur = head

    while (cur !=null){
        if(cur.next != null && cur.val == cur.next.val){
            while(cur.next != null && cur.val == cur.next.val){
                cur = cur.next
            }
            pre.next = cur.next //stretch that pre next pointer all the way to where we've skipped all the dupes
            
        } else {
            pre = cur
        }
        cur = cur.next
    }

    return dummy.next


};